src/main/java/com/viaversion/viafabricplus/protocoltranslator/util/JavaPinger.java

```diff
@@ -0,0 +1,123 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.viaversion.viafabricplus.protocoltranslator.util;
+
+import com.google.gson.Gson;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Very lightweight modern SLP pinger
+ */
+public final class JavaPinger {
+
+    private final static Gson GSON = new Gson();
+
+    public static Response ping(InetSocketAddress address, int timeout, int protocolVersion) {
+        try (Socket socket = new Socket(address.getAddress(), address.getPort())) {
+            socket.setTcpNoDelay(true);
+            socket.setSoTimeout(timeout);
+            socket.connect(null, timeout);
+            try (DataOutputStream dataOutputStream = new DataOutputStream(socket.getOutputStream());
+                 DataInputStream dataInputStream = new DataInputStream(socket.getInputStream());
+                 ByteArrayOutputStream b = new ByteArrayOutputStream();
+                 DataOutputStream handshake = new DataOutputStream(b)) {
+                handshake.writeByte(0);
+                writeVarInt(handshake, protocolVersion);
+                writeVarInt(handshake, address.getHostString().length());
+                handshake.writeBytes(address.getHostString());
+                handshake.writeShort(address.getPort());
+                writeVarInt(handshake, 1);
+                writeVarInt(dataOutputStream, b.size());
+                dataOutputStream.write(b.toByteArray());
+                dataOutputStream.writeByte(1);
+                dataOutputStream.writeByte(0);
+
+                int size = readVarInt(dataInputStream);
+                int id = readVarInt(dataInputStream);
+                int length = readVarInt(dataInputStream);
+
+                if (size <= 0 || id != 0 || length <= 0) {
+                    throw new IllegalStateException("No data received from server");
+                }
+                byte[] in = new byte[length];
+                dataInputStream.readFully(in);
+                return GSON.fromJson(new String(in, StandardCharsets.UTF_8), Response.class);
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException("Could not ping " + address + " with protocol " + protocolVersion, e);
+        }
+    }
+
+    private static int readVarInt(DataInputStream in) throws IOException {
```

Comment:
These already exist in Minecraft

Reply:
Probably. Where can I find these? This code originated from my standalone non-Minecraft mod project.

Reply:
Good thing they haven't changed how to encode VarInt for a decade. Unlikely to change.

Reply:
It would be more about code readability / using the existing classes and methods

Reply:
Looking at `VarInts`, I'd needlessly have to pack everything into a `ByteBuf` and then push it into the raw socket.  I suppose I could do that...

Reply:
No it's fine then I think, that would be too much.